[SPARK-13042][MLlib] Add Specificity (= true negative rate) to MulticlassMetrics.scala

dmarcous opened the pull request:
Add specificity calculations

pecificity (SPC) or true negative rate
SPC = TN / N = TN / (TN+FP) 
False Positive Rate (FPR)
FPR = FP / N = FP / (FP + TN) = 1-SPC

SPC = 1 - FPR

dmarcous committed:
Update MulticlassMetrics.scala

Add specificity calculations

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
There are a number of things wrong with this; please read https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark first. I don't think this is useful enough to include, and your tags are wrong anyway.

dmarcous committed:
Update MulticlassMetricsSuite.scala

Add tests for specificity Metrics

srowen commented:
This is not related to SPARK-10945

dmarcous committed:
Update MulticlassMetrics.scala

changed tags to 1.6.0 

* adding changes to test file

** Thanks for the help, trying to fix everything you points out, as it is a minor change that I think will be helpful for ML people all around

dmarcous committed:
Add specificity calculations

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala:
```diff
@@ -116,6 +116,13 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
+   * Returns specificity for a given label (category)
+   * @param label the label.
+   */
+  @Since("1.6.0")
```
This is still wrong. This can't have existed in 1.6.0

dmarcous committed:
[SPARK-13042][MLlib] Add Specificity calculation tests and tag fixes

dmarcous committed:
[SPARK-13042][MLlib] Add Specificity calculation tests and tag fixes

dmarcous committed:
[SPARK-13042][MLlib] Calculate specificity correctly in test

dmarcous committed:
Merge branch 'master' into patch-1

# Conflicts:
#
mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics
.scala

dmarcous committed:
[SPARK-13042][MLlib] Calculate specificity correctly in test

mengxr commented:
ok to test

SparkQA commented:
**[Test build #51792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51792/consoleFull)** for PR 10945 at commit [`bfc16ea`](https://github.com/apache/spark/commit/bfc16eacfe2f16a009f67bd4a8f6707ec5bc48c2).

SparkQA commented:
**[Test build #51792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51792/consoleFull)** for PR 10945 at commit [`bfc16ea`](https://github.com/apache/spark/commit/bfc16eacfe2f16a009f67bd4a8f6707ec5bc48c2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51792/
Test FAILed.

dmarcous committed:
Update MulticlassMetrics.scala

remove extra whitespaces for scala style tests

dmarcous commented:
Removed extra whitespaces in order to pass style tests

SparkQA commented:
**[Test build #51950 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51950/consoleFull)** for PR 10945 at commit [`7c8f034`](https://github.com/apache/spark/commit/7c8f0346c4be57e674846e94dd40f1efdbecc382).

SparkQA commented:
**[Test build #51950 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51950/consoleFull)** for PR 10945 at commit [`7c8f034`](https://github.com/apache/spark/commit/7c8f0346c4be57e674846e94dd40f1efdbecc382).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51950/
Test FAILed.

dmarcous committed:
[SPARK-13042][MLlib] Correct type

Correct duplicate rows typo

dmarcous commented:
corrected typo - duplicate rows paste

SparkQA commented:
**[Test build #51956 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51956/consoleFull)** for PR 10945 at commit [`f1b2c67`](https://github.com/apache/spark/commit/f1b2c678ddf0c3d408b2b6a39625cbfeed0dc171).

SparkQA commented:
**[Test build #51956 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51956/consoleFull)** for PR 10945 at commit [`f1b2c67`](https://github.com/apache/spark/commit/f1b2c678ddf0c3d408b2b6a39625cbfeed0dc171).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51956/
Test PASSed.

dmarcous commented:
hi guys, any update? seems like its ready for your blessing

srowen commented:
I don't object to someone merging this but I don't think it is worth including in the API. It is trivial to calculate.

dmarcous commented:
My thinking was adding it so it would make code of people using it "cleaner" just specifying another metric by name, and not adding a calculation in user code near the other metrics.

So, if we want to merge and not include in the API does it mean I should mark the new methods as private? remove annotations so it would be undocumented? whats the best way to go from your experience?

srowen commented:
I don't think that would be useful, to make a private unused API.

srowen commented:
I believe we should resolve this as WontFix if there aren't other committers supporting this change

dmarcous commented:
Agreed that it should be resolved.
Is there a process where I can reach out for any committer for support?
I'm still in need of this and think those extra couple of lines goes a long way in user code readability when using it from inside the class

srowen commented:
Paging ... @MLnick ? Usually I find people who have changed surrounding code.

MLnick commented:
I'd only say it's worth putting in if it's something that users would cross-validate against (in which case makes more sense to put it into the `ml` evaluator). But I don't think that's a common case, so unless we saw a lot of user requests I'd tend to lean against it, since as you mention it's pretty straightforward to write a small function to compute it.

dmarcous commented:
Got it, thank you @MLnick and thank you @srowen for your inputs along the way, is has been helpful!
Closing the request.
